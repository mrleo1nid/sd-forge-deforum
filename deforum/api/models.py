# Copyright (C) 2023 Deforum LLC
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, version 3 of the License.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.

# Contact the authors: https://deforum.github.io/

from pydantic import BaseModel, Field, ConfigDict
from typing import Any, Dict, List, Optional, Union
from dataclasses import dataclass
from enum import Enum


class Batch(BaseModel):
    """Request model for submitting a batch of Deforum generation jobs.

    A batch can contain one or multiple Deforum settings objects.
    Each settings object represents a separate video generation job.
    """
    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "deforum_settings": {
                    "animation_mode": "3D",
                    "max_frames": 120,
                    "prompts": {"0": "a beautiful landscape"},
                    "W": 512,
                    "H": 512
                },
                "options_overrides": {
                    "deforum_save_gen_info_as_srt": True
                }
            }
        }
    )

    deforum_settings: Optional[Union[Dict[str, Any], List[Dict[str, Any]]]] = Field(
        default=None,
        description="Single settings object or list of settings objects. "
                    "Each object should match the structure from a saved Deforum settings.txt file."
    )
    options_overrides: Optional[Dict[str, Any]] = Field(
        default=None,
        description="Optional Forge settings to override for this batch (e.g., save_gen_info_as_srt)"
    )

class DeforumJobStatusCategory(str, Enum):
    """Overall status of a Deforum job."""
    ACCEPTED = "ACCEPTED"      # Job accepted and queued
    SUCCEEDED = "SUCCEEDED"    # Job completed successfully
    FAILED = "FAILED"          # Job failed with error
    CANCELLED = "CANCELLED"    # Job cancelled by user


class DeforumJobPhase(str, Enum):
    """Current phase of job execution."""
    QUEUED = "QUEUED"                      # Waiting in queue
    PREPARING = "PREPARING"                # Loading models, preparing args
    GENERATING = "GENERATING"              # Generating frames
    POST_PROCESSING = "POST_PROCESSING"    # Video assembly, upscaling, etc.
    DONE = "DONE"                          # All phases complete


class DeforumJobErrorType(str, Enum):
    """Type of error if job failed."""
    NONE = "NONE"              # No error
    RETRYABLE = "RETRYABLE"    # Temporary error, retry may succeed
    TERMINAL = "TERMINAL"      # Permanent error, retry won't help

class DeforumJobStatus(BaseModel):
    """Detailed status information for a Deforum generation job.

    Tracks the job's progress through various phases and provides
    timing information and output locations.
    """
    model_config = ConfigDict(frozen=True, extra='ignore')

    id: str = Field(description="Unique job identifier")
    status: DeforumJobStatusCategory = Field(description="Overall job status")
    phase: DeforumJobPhase = Field(description="Current execution phase")
    error_type: DeforumJobErrorType = Field(description="Error type if job failed")
    phase_progress: float = Field(
        description="Progress within current phase (0.0 to 1.0)",
        ge=0.0,
        le=1.0
    )
    started_at: float = Field(description="Timestamp when job started (Unix epoch)")
    last_updated: float = Field(description="Timestamp of last status update (Unix epoch)")
    execution_time: float = Field(description="Time elapsed since job start (seconds)")
    update_interval_time: float = Field(description="Time since last status update (seconds)")
    updates: int = Field(description="Number of status updates received", ge=0)
    message: Optional[str] = Field(default=None, description="Error message or status details")
    outdir: Optional[str] = Field(default=None, description="Output directory for generated files")
    timestring: Optional[str] = Field(default=None, description="Unique timestring identifier for output files")
    deforum_settings: Optional[Dict[str, Any]] = Field(
        default=None,
        description="Original Deforum settings used for this job"
    )
    options_overrides: Optional[Dict[str, Any]] = Field(
        default=None,
        description="Forge options overrides used for this job"
    )


class BatchSubmitResponse(BaseModel):
    """Response when a batch is successfully submitted."""
    message: str = Field(description="Success message")
    batch_id: str = Field(description="Unique batch identifier")
    job_ids: List[str] = Field(description="List of job IDs in this batch")


class BatchCancelResponse(BaseModel):
    """Response when cancelling a batch."""
    ids: List[str] = Field(description="List of cancelled job IDs")
    message: str = Field(description="Cancellation result message")


class JobCancelResponse(BaseModel):
    """Response when cancelling a single job."""
    id: str = Field(description="Cancelled job ID")
    message: str = Field(description="Cancellation result message")


class ErrorResponse(BaseModel):
    """Standard error response."""
    id: Optional[str] = Field(default=None, description="Job or batch ID if applicable")
    status: Optional[str] = Field(default=None, description="Status code or category")
    message: str = Field(description="Error message")


class VersionResponse(BaseModel):
    """API or Deforum version information."""
    version: str = Field(description="Version string")


class SimpleRunResponse(BaseModel):
    """Response from the simple /deforum/run endpoint."""
    outdir: str = Field(description="Output directory where generated files will be saved")