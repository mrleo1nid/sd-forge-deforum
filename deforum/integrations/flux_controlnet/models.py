"""Flux ControlNet model loading and management for Deforum.

Handles loading and caching of Flux ControlNet models from HuggingFace.
Supports Canny and Depth ControlNet models.
"""

import torch
from diffusers import FluxControlNetModel, FluxControlNetPipeline
from typing import Optional, Dict
import os
from contextlib import contextmanager


@contextmanager
def temporarily_unpatch_hf_download():
    """Temporarily restore original HuggingFace download to avoid etag parameter conflict.

    Forge patches huggingface_hub download functions with a signature that doesn't
    support the 'etag' parameter that newer diffusers/transformers uses.
    This context manager temporarily restores the original during model loading.
    """
    patched_fn = None
    restored = False

    try:
        from huggingface_hub import file_download

        # Save current (patched) function
        patched_fn = file_download._download_to_tmp_and_move

        # Get the original function from the closure
        # Forge's patch wraps the original: original_download_to_tmp_and_move
        if hasattr(patched_fn, '__code__') and patched_fn.__code__.co_freevars:
            # Try to extract original from closure
            for cell in patched_fn.__closure__ or []:
                try:
                    obj = cell.cell_contents
                    if callable(obj) and obj != patched_fn:
                        # Found the original function
                        file_download._download_to_tmp_and_move = obj
                        restored = True
                        print("  Temporarily using original HF download (avoiding Forge patch)")
                        break
                except (ValueError, AttributeError):
                    continue

        if not restored:
            print("  Using patched HF download (couldn't restore original, may fail)")

    except Exception as e:
        print(f"  Warning during HF download unpatch setup: {e}")

    try:
        yield
    finally:
        # Always restore the patch if we changed it
        if restored and patched_fn is not None:
            try:
                from huggingface_hub import file_download
                file_download._download_to_tmp_and_move = patched_fn
                print("  Restored Forge HF download patch")
            except Exception as e:
                print(f"  Warning: Could not restore HF patch: {e}")


# Available Flux ControlNet models
FLUX_CONTROLNET_MODELS = {
    "canny": {
        "instantx": "InstantX/FLUX.1-dev-Controlnet-Canny",
        "xlabs": "XLabs-AI/flux-controlnet-canny-diffusers",
        "bfl": "black-forest-labs/FLUX.1-Canny-dev",
    },
    "depth": {
        "shakker": "Shakker-Labs/FLUX.1-dev-ControlNet-Depth",
        "instantx": "InstantX/FLUX.1-dev-Controlnet-Depth",
        "xlabs": "XLabs-AI/flux-controlnet-depth-diffusers",
        "bfl": "black-forest-labs/FLUX.1-Depth-dev",
    }
}

# Model cache
_model_cache: Dict[str, FluxControlNetModel] = {}
_pipeline_cache: Optional[FluxControlNetPipeline] = None


def get_available_models(control_type: str) -> Dict[str, str]:
    """Get available models for a control type.

    Args:
        control_type: "canny" or "depth"

    Returns:
        Dictionary of model name -> HuggingFace repo ID
    """
    return FLUX_CONTROLNET_MODELS.get(control_type, {})


def load_flux_controlnet_model(
    control_type: str,
    model_name: str = "instantx",
    torch_dtype: torch.dtype = torch.bfloat16,
    device: str = "cuda"
) -> FluxControlNetModel:
    """Load a Flux ControlNet model.

    Args:
        control_type: "canny" or "depth"
        model_name: Model provider name (e.g., "instantx", "xlabs", "bfl")
        torch_dtype: Torch data type for model weights
        device: Device to load model on

    Returns:
        Loaded FluxControlNetModel

    Raises:
        ValueError: If control_type or model_name is invalid
    """
    # Check if model is already cached
    cache_key = f"{control_type}_{model_name}"
    if cache_key in _model_cache:
        print(f"Using cached Flux {control_type.title()} ControlNet model: {model_name}")
        return _model_cache[cache_key]

    # Get model repo ID
    if control_type not in FLUX_CONTROLNET_MODELS:
        raise ValueError(f"Invalid control type: {control_type}. Use 'canny' or 'depth'.")

    models = FLUX_CONTROLNET_MODELS[control_type]
    if model_name not in models:
        raise ValueError(f"Invalid model name '{model_name}' for {control_type}. "
                        f"Available: {list(models.keys())}")

    model_id = models[model_name]

    print(f"Loading Flux {control_type.title()} ControlNet model: {model_id}")
    print(f"This may take a while on first load (downloading from HuggingFace)...")

    # Load model with temporarily unpatched HF download to avoid etag parameter conflict
    try:
        with temporarily_unpatch_hf_download():
            controlnet = FluxControlNetModel.from_pretrained(
                model_id,
                torch_dtype=torch_dtype
            )

        # Cache the model
        _model_cache[cache_key] = controlnet

        print(f"✓ Flux {control_type.title()} ControlNet model loaded successfully")
        return controlnet

    except Exception as e:
        print(f"Error loading Flux ControlNet model {model_id}: {e}")
        raise


def load_flux_controlnet_pipeline(
    base_model: str = "black-forest-labs/FLUX.1-dev",
    controlnet: Optional[FluxControlNetModel] = None,
    control_type: str = "canny",
    model_name: str = "instantx",
    torch_dtype: torch.dtype = torch.bfloat16,
    device: str = "cuda"
) -> FluxControlNetPipeline:
    """Load or create a Flux ControlNet pipeline.

    Args:
        base_model: Base Flux model repo ID
        controlnet: Pre-loaded ControlNet model (optional)
        control_type: "canny" or "depth" (if controlnet not provided)
        model_name: Model provider name (if controlnet not provided)
        torch_dtype: Torch data type
        device: Device to load on

    Returns:
        FluxControlNetPipeline ready for inference
    """
    global _pipeline_cache

    # Load ControlNet if not provided
    if controlnet is None:
        controlnet = load_flux_controlnet_model(control_type, model_name, torch_dtype, device)

    # Check if we can reuse cached pipeline (for now, always create new)
    # TODO: Implement pipeline caching with hot-swappable ControlNet

    print(f"Creating Flux ControlNet pipeline with base model: {base_model}")

    try:
        with temporarily_unpatch_hf_download():
            pipe = FluxControlNetPipeline.from_pretrained(
                base_model,
                controlnet=controlnet,
                torch_dtype=torch_dtype
            )
        pipe.to(device)

        print("✓ Flux ControlNet pipeline created successfully")
        return pipe

    except Exception as e:
        print(f"Error creating Flux ControlNet pipeline: {e}")
        raise


def unload_controlnet_models():
    """Clear cached models to free VRAM."""
    global _model_cache, _pipeline_cache

    _model_cache.clear()
    _pipeline_cache = None

    if torch.cuda.is_available():
        torch.cuda.empty_cache()

    print("Flux ControlNet models unloaded from cache")


def get_model_info(control_type: str, model_name: str) -> str:
    """Get human-readable info about a model.

    Args:
        control_type: "canny" or "depth"
        model_name: Model provider name

    Returns:
        Model information string
    """
    models = FLUX_CONTROLNET_MODELS.get(control_type, {})
    repo_id = models.get(model_name, "Unknown")

    provider_names = {
        "instantx": "InstantX",
        "xlabs": "XLabs-AI",
        "bfl": "Black Forest Labs (Official)",
        "shakker": "Shakker Labs"
    }

    provider = provider_names.get(model_name, model_name)

    return f"{control_type.title()} ControlNet by {provider} ({repo_id})"
