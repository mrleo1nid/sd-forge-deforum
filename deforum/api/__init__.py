"""Deforum REST API module.

This module provides REST API endpoints for running Deforum generations
and managing job status through FastAPI integration.

API endpoints are automatically registered via script_callbacks.on_app_started()
when the module is imported (see bottom of api.py).
"""

from .api import JobStatusTracker
from .models import (
    Batch,
    DeforumJobErrorType,
    DeforumJobPhase,
    DeforumJobStatus,
    DeforumJobStatusCategory,
)

__all__ = [
    "JobStatusTracker",
    "Batch",
    "DeforumJobErrorType",
    "DeforumJobPhase",
    "DeforumJobStatus",
    "DeforumJobStatusCategory",
]
