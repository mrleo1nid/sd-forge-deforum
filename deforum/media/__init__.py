"""Media handling module for Deforum.

Provides image/video I/O, upscaling, and interpolation capabilities.
"""

from .load_images import (
    load_img,
    load_image,
    get_mask,
    get_mask_from_file,
    # Add other load functions as needed
)
from .save_images import (
    save_image,
    # Add other save functions as needed
)
from .upscaling import (
    process_ncnn_upscale_vid_upload_logic,
    process_ncnn_video_upscaling,
    make_upscale_v2,
)
from .video_audio_utilities import (
    get_frame_name,
    get_next_frame,
    render_preview,
    download_audio,
    # Add other video/audio functions as needed
)

__all__ = [
    # Load
    "load_img",
    "load_image",
    "get_mask",
    "get_mask_from_file",
    # Save
    "save_image",
    # Upscaling
    "process_ncnn_upscale_vid_upload_logic",
    "process_ncnn_video_upscaling",
    "make_upscale_v2",
    # Video/Audio
    "get_frame_name",
    "get_next_frame",
    "render_preview",
    "download_audio",
]
