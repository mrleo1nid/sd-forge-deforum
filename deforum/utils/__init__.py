"""
Deforum utility modules - pure functions organized by domain.

This package contains refactored pure functions extracted from various
legacy Deforum modules in scripts/deforum_helpers/, organized by functional
domain for better maintainability and testability.

All functions in this package follow functional programming principles:
- Pure functions with no side effects
- Complete type hints
- Comprehensive docstrings
- Unit tested with >70% coverage

Modules:
    seed_utils: Seed generation and iteration logic
    image_utils: Image processing (sharpening, color matching, conversions)
    noise_utils: Perlin noise generation
    prompt_utils: Prompt parsing and interpolation
    transform_utils: 3D transformations and matrix operations
    functional_utils: Functional programming helpers and utilities
    filename_utils: Filename formatting and path generation
    subtitle_utils: Subtitle time formatting and parameter display
    string_utils: String manipulation and formatting utilities
    path_utils: Path manipulation and parsing utilities
    validation_utils: Image validation and checking utilities
    math_utils: Mathematical calculations and 3D transformations
    file_utils: File and path operations
    optical_flow_utils: Optical flow consistency checking
    format_utils: Value formatting and conversion utilities
    interpolation_utils: Frame interpolation calculations
    video_path_utils: Video and image path generation
"""

__all__ = [
    "seed_utils",
    "image_utils",
    "noise_utils",
    "prompt_utils",
    "transform_utils",
    "functional_utils",
    "filename_utils",
    "subtitle_utils",
    "string_utils",
    "path_utils",
    "validation_utils",
    "math_utils",
    "file_utils",
    "optical_flow_utils",
    "format_utils",
    "interpolation_utils",
    "video_path_utils",
]
