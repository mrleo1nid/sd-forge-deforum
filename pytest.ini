[pytest]
filterwarnings =
    ignore::DeprecationWarning

# Test markers
markers =
    slow: mark test as slow (generates real outputs)
    flux: mark test as requiring Flux model
    wan: mark test as requiring Wan model
    visual: mark test as testing visual processing features

# Coverage configuration
addopts =
    --cov=scripts/deforum_helpers
    --cov=scripts/deforum_api.py
    --cov=scripts/deforum_api_models.py
    --cov-report=html
    --cov-report=term-missing
    --cov-branch

# Coverage options
[coverage:run]
source = scripts
omit =
    */tests/*
    */venv/*
    */__pycache__/*
    */site-packages/*

[coverage:report]
# Fail if coverage drops below this threshold (disabled for now)
# fail_under = 70
exclude_lines =
    pragma: no cover
    def __repr__
    raise AssertionError
    raise NotImplementedError
    if __name__ == .__main__.:
    if TYPE_CHECKING:
    @abstractmethod
