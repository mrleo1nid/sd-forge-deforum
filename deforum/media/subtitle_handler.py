# Copyright (C) 2023 Deforum LLC
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, version 3 of the License.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.

# Contact the authors: https://deforum.github.io/

"""Subtitle handling - Mixed pure and impure functions.

This module contains:
- Pure formatting/calculation functions imported from deforum.utils.subtitle_utils
- Impure I/O functions for writing subtitle files (kept here due to side effects)
"""

from decimal import Decimal

# Import pure functions from refactored utils module
from deforum.utils.subtitle_utils import (
    time_to_srt_format,
    calculate_frame_duration,
    frame_time,
    format_subtitle_value as _format_value,
    SUBTITLE_PARAM_NAMES as param_dict,
    get_user_param_names,
)

# Backward compatibility alias
get_user_values = get_user_param_names


def init_srt_file(filename, fps, precision=20):
    with open(filename, "w") as f:
        pass
    return calculate_frame_duration(fps)


def write_frame_subtitle(filename, frame_number, frame_duration, text):
    # Used by stable core only. Meant to be used when subtitles are intended to change with every frame.
    frames_per_subtitle = 1
    # For low FPS animations and for debugging this is fine, but at higher FPS the file may be too bloated and not fit
    # for YouTube upload, in which case "write_subtitle_from_to" may be used directly with a longer duration instead.
    start_time = frame_time(frame_number, frame_duration)
    end_time = (Decimal(frame_number) + Decimal(frames_per_subtitle)) * frame_duration
    write_subtitle_from_to(filename, frame_number, start_time, end_time, text)


def write_subtitle_from_to(filename, frame_number, start_time_s, end_time_s, text):
    # start_time should be the same as end_time of the last call
    with open(filename, "a", encoding="utf-8") as f:
        # see https://en.wikipedia.org/wiki/SubRip#Format
        f.write(f"{frame_number + 1}\n")
        f.write(f"{time_to_srt_format(start_time_s)} --> {time_to_srt_format(end_time_s)}\n")
        f.write(f"{text}\n\n")


def format_animation_params(keys, prompt_series, frame_idx, params_to_print):
    params_string = ""
    for key, value in param_dict.items():
        if value['user'] in params_to_print:
            backend_key = value['backend']
            print_key = value['print']
            param_value = getattr(keys, backend_key)[frame_idx]
            formatted_value = _format_value(param_value)
            params_string += f"{print_key}: {formatted_value}; "

    if "Prompt" in params_to_print:
        params_string += f"Prompt: {prompt_series[frame_idx]}; "

    params_string = params_string.rstrip("; ")  # Remove trailing semicolon and whitespace
    return params_string
