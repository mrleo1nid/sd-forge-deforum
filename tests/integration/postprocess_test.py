# Copyright (C) 2023 Deforum LLC
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, version 3 of the License.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.

# Contact the authors: https://deforum.github.io/

import glob
import json
import os
from pathlib import Path

import pytest
import requests
from moviepy.editor import VideoFileClip
from .utils import (API_BASE_URL, get_test_options_overrides, gpu_disabled, wait_for_job_to_complete, get_test_batch_name)

from deforum.api.models import (DeforumJobPhase,
                                        DeforumJobStatusCategory)
from deforum.media.subtitle_handler import get_user_values


# Path to testdata directory relative to this test file
TESTDATA_DIR = Path(__file__).parent / 'testdata'


@pytest.mark.skipif(gpu_disabled(), reason="requires GPU-enabled server")
def test_post_process_FILM(snapshot):
    with open(TESTDATA_DIR / 'simple.input_settings.txt', 'r') as settings_file:
        deforum_settings = json.load(settings_file)

    # Set test-specific batch name for easier output identification
    deforum_settings['batch_name'] = get_test_batch_name('test_post_process_FILM')

    deforum_settings["frame_interpolation_engine"] = "FILM"
    deforum_settings["frame_interpolation_x_amount"] = 3
    deforum_settings["frame_interpolation_slow_mo_enabled"] = False

    options_overrides = get_test_options_overrides()
    options_overrides.update({
        "deforum_save_gen_info_as_srt": True,
        "deforum_save_gen_info_as_srt_params": get_user_values(),
    })

    response = requests.post(API_BASE_URL+"/batches", json={
        "deforum_settings":[deforum_settings],
        "options_overrides": options_overrides
        })
    response.raise_for_status()
    job_id = response.json()["job_ids"][0]
    jobStatus = wait_for_job_to_complete(job_id)

    assert jobStatus.status == DeforumJobStatusCategory.SUCCEEDED, f"Job {job_id} failed: {jobStatus}"

    # Ensure parameters used at each frame have not regressed
    srt_filenname =  os.path.join(jobStatus.outdir, f"{jobStatus.timestring}.srt")
    with open(srt_filenname, 'r') as srt_file:
        assert srt_file.read() == snapshot

    # Ensure interpolated video format is as expected
    video_filenames = glob.glob(f'{jobStatus.outdir}/*FILM*.mp4', recursive=True)
    assert len(video_filenames) == 1, "Expected one FILM video to be generated"

    interpolated_video_filename =  video_filenames[0]
    clip = VideoFileClip(interpolated_video_filename)
    assert clip.fps == deforum_settings['fps'] * deforum_settings["frame_interpolation_x_amount"]  , "Video FPS does not match input settings (fps * interpolation amount)"
    # FILM interpolation: N frames + (N-1) gaps × (x-1) interpolated frames per gap
    expected_frame_count = deforum_settings['max_frames'] + (deforum_settings['max_frames'] - 1) * (deforum_settings["frame_interpolation_x_amount"] - 1)
    assert clip.duration * clip.fps == expected_frame_count, f"Video frame count does not match FILM interpolation formula (expected {expected_frame_count})"
    assert clip.size == [deforum_settings['W'], deforum_settings['H']] , "Video dimensions are not as expected"    

@pytest.mark.skipif(gpu_disabled(), reason="requires GPU-enabled server")
def test_post_process_RIFE(snapshot):
    with open(TESTDATA_DIR / 'simple.input_settings.txt', 'r') as settings_file:
        deforum_settings = json.load(settings_file)

    # Set test-specific batch name for easier output identification
    deforum_settings['batch_name'] = get_test_batch_name('test_post_process_RIFE')

    deforum_settings["frame_interpolation_engine"] = "RIFE v4.6"
    deforum_settings["frame_interpolation_x_amount"] = 3
    deforum_settings["frame_interpolation_slow_mo_enabled"] = False

    options_overrides = get_test_options_overrides()
    options_overrides.update({
        "deforum_save_gen_info_as_srt": True,
        "deforum_save_gen_info_as_srt_params": get_user_values(),
    })

    response = requests.post(API_BASE_URL+"/batches", json={
        "deforum_settings":[deforum_settings],
        "options_overrides": options_overrides
        })
    response.raise_for_status()
    job_id = response.json()["job_ids"][0]
    jobStatus = wait_for_job_to_complete(job_id)

    assert jobStatus.status == DeforumJobStatusCategory.SUCCEEDED, f"Job {job_id} failed: {jobStatus}"

    # Ensure parameters used at each frame have not regressed
    srt_filenname =  os.path.join(jobStatus.outdir, f"{jobStatus.timestring}.srt")
    with open(srt_filenname, 'r') as srt_file:
        assert srt_file.read() == snapshot

    # Ensure interpolated video format is as expected
    video_filenames = glob.glob(f'{jobStatus.outdir}/*RIFE*.mp4', recursive=True)
    assert len(video_filenames) == 1, "Expected one RIFE video to be generated"

    interpolated_video_filename =  video_filenames[0]
    clip = VideoFileClip(interpolated_video_filename)
    assert clip.fps == deforum_settings['fps'] * deforum_settings["frame_interpolation_x_amount"]  , "Video FPS does not match input settings (fps * interpolation amount)"
    # RIFE interpolation: N frames + (N-1) gaps × (x-1) interpolated frames per gap
    expected_frame_count = deforum_settings['max_frames'] + (deforum_settings['max_frames'] - 1) * (deforum_settings["frame_interpolation_x_amount"] - 1)
    assert clip.duration * clip.fps == expected_frame_count, f"Video frame count does not match RIFE interpolation formula (expected {expected_frame_count})"
    assert clip.size == [deforum_settings['W'], deforum_settings['H']] , "Video dimensions are not as expected"        

@pytest.mark.skipif(gpu_disabled(), reason="requires GPU-enabled server")
def test_post_process_UPSCALE(snapshot):
    with open(TESTDATA_DIR / 'simple.input_settings.txt', 'r') as settings_file:
        deforum_settings = json.load(settings_file)

    # Set test-specific batch name for easier output identification
    deforum_settings['batch_name'] = get_test_batch_name('test_post_process_UPSCALE')

    deforum_settings["r_upscale_video"] = True
    deforum_settings["r_upscale_factor"] = "x4"
    deforum_settings["r_upscale_model"] = "realesrgan-x4plus"

    options_overrides = get_test_options_overrides()
    options_overrides.update({
        "deforum_save_gen_info_as_srt": True,
        "deforum_save_gen_info_as_srt_params": get_user_values(),
    })

    response = requests.post(API_BASE_URL+"/batches", json={
        "deforum_settings":[deforum_settings],
        "options_overrides": options_overrides
        })
    response.raise_for_status()
    job_id = response.json()["job_ids"][0]
    jobStatus = wait_for_job_to_complete(job_id)

    assert jobStatus.status == DeforumJobStatusCategory.SUCCEEDED, f"Job {job_id} failed: {jobStatus}"

    # Ensure parameters used at each frame have not regressed
    srt_filenname =  os.path.join(jobStatus.outdir, f"{jobStatus.timestring}.srt")
    with open(srt_filenname, 'r') as srt_file:
        assert srt_file.read() == snapshot

    # Ensure interpolated video format is as expected
    video_filenames = glob.glob(f'{jobStatus.outdir}/*Upscaled*.mp4', recursive=True)
    assert len(video_filenames) == 1, "Expected one upscaled video to be generated"

    interpolated_video_filename =  video_filenames[0]
    clip = VideoFileClip(interpolated_video_filename)
    assert clip.fps == deforum_settings['fps'] , "Video FPS does not match input settings"
    assert clip.duration * clip.fps == deforum_settings['max_frames'], "Video frame count does not match input settings"
    assert clip.size == [deforum_settings['W']*4, deforum_settings['H']*4] , "Video dimensions are not as expected (including upscaling)"


@pytest.mark.skipif(gpu_disabled(), reason="requires GPU-enabled server")
def test_post_process_UPSCALE_FILM(snapshot):
    with open(TESTDATA_DIR / 'simple.input_settings.txt', 'r') as settings_file:
        deforum_settings = json.load(settings_file)

    # Set test-specific batch name for easier output identification
    deforum_settings['batch_name'] = get_test_batch_name('test_post_process_UPSCALE_FILM')

    deforum_settings["r_upscale_video"] = True
    deforum_settings["r_upscale_factor"] = "x4"
    deforum_settings["r_upscale_model"] = "realesrgan-x4plus"
    deforum_settings["frame_interpolation_engine"] = "FILM"
    deforum_settings["frame_interpolation_x_amount"] = 3
    deforum_settings["frame_interpolation_slow_mo_enabled"] = False
    deforum_settings["frame_interpolation_use_upscaled"] = True

    options_overrides = get_test_options_overrides()
    options_overrides.update({
        "deforum_save_gen_info_as_srt": True,
        "deforum_save_gen_info_as_srt_params": get_user_values(),
    })

    response = requests.post(API_BASE_URL+"/batches", json={
        "deforum_settings":[deforum_settings],
        "options_overrides": options_overrides
        })
    response.raise_for_status()
    job_id = response.json()["job_ids"][0]
    jobStatus = wait_for_job_to_complete(job_id)

    assert jobStatus.status == DeforumJobStatusCategory.SUCCEEDED, f"Job {job_id} failed: {jobStatus}"

    # Ensure parameters used at each frame have not regressed
    srt_filenname =  os.path.join(jobStatus.outdir, f"{jobStatus.timestring}.srt")
    with open(srt_filenname, 'r') as srt_file:
        assert srt_file.read() == snapshot

    # Ensure interpolated video format is as expected
    video_filenames = glob.glob(f'{jobStatus.outdir}/*FILM*upscaled*.mp4', recursive=True)
    assert len(video_filenames) == 1, "Expected one upscaled video to be generated"

    interpolated_video_filename =  video_filenames[0]
    clip = VideoFileClip(interpolated_video_filename)
    assert clip.fps == deforum_settings['fps'] * deforum_settings["frame_interpolation_x_amount"]  , "Video FPS does not match input settings (fps * interpolation amount)"
    # FILM may use different interpolation strategy - check actual frame count
    # For now, use same formula as RIFE: N frames + (N-1) gaps × (x-1) interpolated frames per gap
    expected_frame_count = deforum_settings['max_frames'] + (deforum_settings['max_frames'] - 1) * (deforum_settings["frame_interpolation_x_amount"] - 1)
    assert clip.duration * clip.fps == expected_frame_count, f"Video frame count does not match interpolation formula (expected {expected_frame_count})"
    assert clip.size == [deforum_settings['W']*4, deforum_settings['H']*4] , "Video dimensions are not as expected (including upscaling)"    
