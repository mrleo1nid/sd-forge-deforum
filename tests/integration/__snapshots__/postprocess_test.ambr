# serializer version: 1
# name: test_post_process_FILM
  '''
  1
  00:00:00,000 --> 00:00:00,050
  F#: 00001; Prompt:  tiny cute swamp bunny, highly detailed, intricate, ultra hd, sharp photo, crepuscular rays, in focus
  
  2
  00:00:00,050 --> 00:00:00,550
  F#: 00011; Prompt:  tiny cute swamp bunny, highly detailed, intricate, ultra hd, sharp photo, crepuscular rays, in focus
  
  
  '''
# ---
# name: test_post_process_RIFE
  '''
  1
  00:00:00,000 --> 00:00:00,050
  F#: 00001; Prompt:  tiny cute swamp bunny, highly detailed, intricate, ultra hd, sharp photo, crepuscular rays, in focus
  
  2
  00:00:00,050 --> 00:00:00,550
  F#: 00011; Prompt:  tiny cute swamp bunny, highly detailed, intricate, ultra hd, sharp photo, crepuscular rays, in focus
  
  
  '''
# ---
# name: test_post_process_UPSCALE
  '''
  1
  00:00:00,000 --> 00:00:00,050
  F#: 00001; Prompt:  tiny cute swamp bunny, highly detailed, intricate, ultra hd, sharp photo, crepuscular rays, in focus
  
  2
  00:00:00,050 --> 00:00:00,550
  F#: 00011; Prompt:  tiny cute swamp bunny, highly detailed, intricate, ultra hd, sharp photo, crepuscular rays, in focus
  
  
  '''
# ---
